File: FivesLabJAVA/src/FirstZ.java
/*
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

// Класс, представляющий ячейку склада
class Cell {
    String address; // Адрес ячейки
    Map<String, Integer> products; // Карта товаров и их количества в ячейке

    // Конструктор для создания новой ячейки
    public Cell(String addr) {
        address = addr;
        products = new HashMap<>(); // Инициализация карты товаров
    }

    // Метод для добавления товара в ячейку
    public void addProduct(String productName, int quantity) {
        products.put(productName, products.getOrDefault(productName, 0) + quantity); // Добавление товара в карту
        System.out.println("Добавлено " + quantity + " единиц товара " + productName + " в ячейку " + address + ".");
    }

    // Метод для удаления товара из ячейки
    public void removeProduct(String productName, int quantity) {
        if (products.containsKey(productName)) { // Проверка наличия товара в ячейке
            int currentQuantity = products.get(productName);
            if (currentQuantity >= quantity) { // Проверка достаточного количества товара для удаления
                products.put(productName, currentQuantity - quantity); // Удаление товара из карты
                if (products.get(productName) == 0) { // Если количество товара стало нулевым, удаляем его из карты
                    products.remove(productName);
                }
                System.out.println("Убрано " + quantity + " единиц товара " + productName + " из ячейки " + address + ".");
            } else {
                System.out.println("Невозможно убрать " + quantity + " единиц товара " + productName + " из ячейки " + address + ".");
            }
        } else {
            System.out.println("Невозможно убрать " + quantity + " единиц товара " + productName + " из ячейки " + address + ".");
        }
    }

    // Метод для вывода содержимого ячейки
    public void showContents() {
        if (!products.isEmpty()) { // Проверка наличия товаров в ячейке
            System.out.println("Содержимое ячейки " + address + ":");
            for (Map.Entry<String, Integer> entry : products.entrySet()) {
                System.out.println("- " + entry.getKey() + ": " + entry.getValue() + " шт.");
            }
        }
    }
}

// Класс, представляющий склад
class Warehouse {
    int zones; // Количество зон на складе
    int shelvesPerZone; // Количество полок в каждой зоне
    int sectionsPerShelf; // Количество секций на каждой полке
    int shelvesCapacity; // Вместимость каждой полки
    int totalCapacity; // Общая вместимость склада
    Map<String, Cell> cells; // Карта ячеек склада

    // Конструктор для создания нового склада с заданными параметрами
    public Warehouse(int z, int spz, int sps, int sc) {
        zones = z;
        shelvesPerZone = spz;
        sectionsPerShelf = sps;
        shelvesCapacity = sc;
        totalCapacity = zones * shelvesPerZone * sectionsPerShelf * shelvesCapacity; // Вычисление общей вместимости
        cells = new HashMap<>(); // Инициализация карты ячеек

        // Создание ячеек для каждой комбинации зоны, полки, секции и позиции на полке
        for (int zone = 1; zone <= zones; ++zone) {
            for (int shelf = 1; shelf <= shelvesPerZone; ++shelf) {
                for (int section = 1; section <= sectionsPerShelf; ++section) {
                    for (int shelfPosition = 1; shelfPosition <= shelvesCapacity; ++shelfPosition) {
                        String address = "" + (char) ('A' + zone - 1) + shelf + section + shelfPosition; // Генерация адреса
                        cells.put(address, new Cell(address)); // Добавление новой ячейки в карту
                    }
                }
            }
        }
    }

    // Метод для добавления товара в указанную ячейку
    public void addProduct(String productName, int quantity, String cellAddress) {
        Cell cell = cells.get(cellAddress); // Получение ячейки по адресу
        if (cell != null) {
            cell.addProduct(productName, quantity); // Добавление товара в ячейку
        } else {
            System.out.println("Ячейка не существует.");
        }
    }

    // Метод для удаления товара из указанной ячейки
    public void removeProduct(String productName, int quantity, String cellAddress) {
        Cell cell = cells.get(cellAddress); // Получение ячейки по адресу
        if (cell != null) {
            cell.removeProduct(productName, quantity); // Удаление товара из ячейки
        } else {
            System.out.println("Ячейка не существует.");
        }
    }

    // Метод для вывода информации о заполненных ячейках склада
    public void getInfo() {
        int totalFilledCapacity = 0; // Общее количество заполненных ячеек
        for (Cell cell : cells.values()) {
            if (!cell.products.isEmpty()) { // Проверка наличия товаров в ячейке
                totalFilledCapacity += cell.products.size(); // Увеличение счетчика заполненных ячеек
                cell.showContents(); // Вывод содержимого ячейки
            }
        }
        double filledPercentage = ((double) totalFilledCapacity / totalCapacity) * 100.0; // Вычисление процента заполненности
        System.out.println("Склад заполнен на " + filledPercentage + "%.");
    }

    // Метод для вывода списка пустых ячеек
    public void showEmptyCells() {
        System.out.println("Пустые ячейки:");
        for (Cell cell : cells.values()) {
            if (cell.products.isEmpty()) { // Проверка пустоты ячейки
                System.out.println(cell.address); // Вывод адреса пустой ячейки
            }
        }
    }
}

// Основной класс программы
public class FirstZ {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Warehouse warehouse = new Warehouse(2, 19, 4, 8); // Создание склада с заданными параметрами

        // Вывод инструкции по использованию программы
        System.out.println("Для добавления товара в ячейку используйте команду ADD <наименование товара> <количество> <адрес ячейки>.");
        System.out.println("Пример: ADD Апельсины 8 А1739");
        System.out.println("Для удаления товара из ячейки используйте команду REMOVE <наименование товара> <количество> <адрес ячейки>.");
        System.out.println("Пример: REMOVE Апельсины 3 А1739");
        System.out.println("Для получения информации о состоянии заполненных ячеек склада используйте команду INFO.");
        System.out.println("Для получения списка пустых ячеек используйте команду EMPTY.");
        System.out.println("Команда HELP - выводит инструкцию по вводу.");
        System.out.println("Команда EXIT - завершает программу.");

        // Цикл обработки команд пользователя
        while (true) {
            System.out.print("Введите команду: ");
            String command = scanner.nextLine();

            if (command.equals("EXIT")) { // Проверка команды на завершение программы
                break;
            } else if (command.equals("HELP")) { // Вывод справочной информации
                System.out.println("Для добавления товара в ячейку используйте команду ADD <наименование товара> <количество> <адрес ячейки>.");
                System.out.println("Пример: ADD Апельсины 8 А1739");
                System.out.println("Для удаления товара из ячейки используйте команду REMOVE <наименование товара> <количество> <адрес ячейки>.");
                System.out.println("Пример: REMOVE Апельсины 3 А1739");
                System.out.println("Для получения информации о состоянии заполненных ячеек склада используйте команду INFO.");
                System.out.println("Для получения списка пустых ячеек используйте команду EMPTY.");
                System.out.println("Команда HELP - выводит инструкцию по вводу.");
                System.out.println("Команда EXIT - завершает программу.");
            } else {
                if (command.startsWith("ADD")) { // Обработка команды добавления товара
                    String[] parts = command.split(" ");
                    if (parts.length == 4) {
                        String productName = parts[1];
                        int quantity = Integer.parseInt(parts[2]);
                        String cellAddress = parts[3];
                        warehouse.addProduct(productName, quantity, cellAddress); // Вызов метода добавления товара на склад
                    } else {
                        System.out.println("Некорректная команда.");
                    }
                } else if (command.startsWith("REMOVE")) { // Обработка команды удаления товара
                    String[] parts = command.split(" ");
                    if (parts.length == 4) {
                        String productName = parts[1];
                        int quantity = Integer.parseInt(parts[2]);
                        String cellAddress = parts[3];
                        warehouse.removeProduct(productName, quantity, cellAddress); // Вызов метода удаления товара со склада
                    } else {
                        System.out.println("Некорректная команда.");
                    }
                } else if (command.equals("INFO")) { // Обработка команды вывода информации о складе
                    warehouse.getInfo(); // Вызов метода вывода информации о складе
                } else if (command.equals("EMPTY")) { // Обработка команды вывода списка пустых ячеек
                    warehouse.showEmptyCells(); // Вызов метода вывода списка пустых ячеек
                } else {
                    System.out.println("Неизвестная команда.");
                }
            }
        }

        scanner.close();
    }
}
*/
import java.util.*;

class Cell {
    String address;
    Map<String, Integer> products;

    Cell(String addr) {
        address = addr;
        products = new HashMap<>();
    }

    void addProduct(String productName, int quantity) {
        if (products.containsKey(productName)) {
            if (products.get(productName) + quantity <= 10 && quantity > 0) {
                products.put(productName, products.get(productName) + quantity);
                System.out.println("Добавлено " + quantity + " единиц товара " + productName + " в ячейку " + address + ".");
            } else if (quantity <= 0) {
                System.out.println("Нельзя добавить неположительное количество товара (" + productName + ") в ячейку " + address + ".");
            } else {
                System.out.println("Превышено максимальное количество товара (" + productName + ") в ячейке " + address + ".");
            }
        } else {
            int totalQuantity = 0;
            for (int value : products.values()) {
                totalQuantity += value;
            }
            if (totalQuantity + quantity <= 10 && totalQuantity + quantity > 0) {
                products.put(productName, quantity);
                System.out.println("Добавлено " + quantity + " единиц товара " + productName + " в ячейку " + address + ".");
            } else if (totalQuantity + quantity <= 0) {
                System.out.println("Нельзя добавить неположительное количество товара (" + productName + ") в ячейку " + address + ".");
            } else {
                System.out.println("Превышено максимальное количество товара (" + productName + ") в ячейке " + address + ".");
            }
        }
    }

    void removeProduct(String productName, int quantity) {
        if (products.containsKey(productName) && products.get(productName) >= quantity && quantity >= 0) {
            products.put(productName, products.get(productName) - quantity);
            if (products.get(productName) == 0) {
                products.remove(productName);
            }
            System.out.println("Убрано " + quantity + " единиц товара " + productName + " из ячейки " + address + ".");
        } else {
            System.out.println("Невозможно убрать " + quantity + " единиц товара " + productName + " из ячейки " + address + ".");
        }
    }

    void showContents() {
        if (!products.isEmpty()) {
            System.out.println("Содержимое ячейки " + address + ":");
            for (Map.Entry<String, Integer> entry : products.entrySet()) {
                System.out.println("- " + entry.getKey() + ": " + entry.getValue() + " шт.");
            }
        }
    }
}

class Warehouse {
    int zones;
    int shelvesPerZone;
    int sectionsPerShelf;
    int shelvesCapacity;
    int totalCapacity;
    Map<String, Cell> cells;

    Warehouse(int z, int spz, int sps, int sc) {
        zones = z;
        shelvesPerZone = spz;
        sectionsPerShelf = sps;
        shelvesCapacity = sc;
        totalCapacity = zones * shelvesPerZone * sectionsPerShelf * shelvesCapacity;
        cells = new HashMap<>();
        for (int zone = 1; zone <= zones; ++zone) {
            for (int shelf = 1; shelf <= shelvesPerZone; ++shelf) {
                for (int section = 1; section <= sectionsPerShelf; ++section) {
                    for (int shelfPosition = 1; shelfPosition <= shelvesCapacity; ++shelfPosition) {
                        String address = "" + (char)('A' + zone - 1) + shelf + section + shelfPosition;
                        cells.put(address, new Cell(address));
                    }
                }
            }
        }
    }

    void addProduct(String productName, int quantity, String cellAddress) {
        Cell cell = cells.get(cellAddress);
        if (cell == null) {
            System.out.println("Ячейка не существует.");
            return;
        }
        cell.addProduct(productName, quantity);
    }

    void removeProduct(String productName, int quantity, String cellAddress) {
        Cell cell = cells.get(cellAddress);
        if (cell == null) {
            System.out.println("Ячейка не существует.");
            return;
        }
        cell.removeProduct(productName, quantity);
    }

    void getInfo() {
        List<Cell> filledCells = new ArrayList<>();
        List<String> emptyCells = new ArrayList<>();
        int totalFilledCapacity = 0;
        for (Cell cell : cells.values()) {
            if (!cell.products.isEmpty()) {
                filledCells.add(cell);
                totalFilledCapacity += cell.products.size();
            } else {
                emptyCells.add(cell.address);
            }
        }
        double filledPercentage = ((double) totalFilledCapacity / totalCapacity) * 100.0;
        System.out.println("Склад заполнен на " + filledPercentage + "%.");
        System.out.println("Содержимое заполненных ячеек:");
        for (Cell cell : filledCells) {
            cell.showContents();
        }
    }

    void showEmptyCells() {
        System.out.println("Пустые ячейки:");
        for (Cell cell : cells.values()) {
            if (cell.products.isEmpty()) {
                System.out.println(cell.address);
            }
        }
    }
}

public class FirstZ {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Warehouse warehouse = new Warehouse(2, 19, 4, 8);

        System.out.println("Для добавления товара в ячейку используйте команду ADD <наименование товара> <количество> <адрес ячейки>.");
        System.out.println("Пример: ADD Апельсины 8 А1739");
        System.out.println("Для удаления товара из ячейки используйте команду REMOVE <наименование товара> <количество> <адрес ячейки>.");
        System.out.println("Пример: REMOVE Апельсины 3 А1739");
        System.out.println("Для получения информации о состоянии заполненных ячеек склада используйте команду INFO.");
        System.out.println("Для получения списка пустых ячеек используйте команду EMPTY.");
        System.out.println("Команда HELP - выводит инструкцию по вводу.");
        System.out.println("Команда EXIT - завершает программу.");

        while (true) {
            System.out.print(">>> ");
            String command = scanner.nextLine();

            if (command.equals("EXIT")) {
                break;
            } else if (command.equals("HELP")) {
                System.out.println("Для добавления товара в ячейку используйте команду ADD <наименование товара> <количество> <адрес ячейки>.");
                System.out.println("Пример: ADD Апельсины 8 А1739");
                System.out.println("Для удаления товара из ячейки используйте команду REMOVE <наименование товара> <количество> <адрес ячейки>.");
                System.out.println("Пример: REMOVE Апельсины 3 А1739");
                System.out.println("Для получения информации о состоянии заполненных ячеек склада используйте команду INFO.");
                System.out.println("Для получения списка пустых ячеек используйте команду EMPTY.");
                System.out.println("Команда HELP - выводит инструкцию по вводу.");
                System.out.println("Команда EXIT - завершает программу.");
            } else if (command.startsWith("ADD")) {
                String[] parts = command.split(" ");
                if (parts.length == 4) {
                    String productName = parts[1];
                    int quantity = Integer.parseInt(parts[2]);
                    String cellAddress = parts[3];
                    warehouse.addProduct(productName, quantity, cellAddress);
                } else {
                    System.out.println("Неверный формат команды.");
                }
            } else if (command.startsWith("REMOVE")) {
                String[] parts = command.split(" ");
                if (parts.length == 4) {
                    String productName = parts[1];
                    int quantity = Integer.parseInt(parts[2]);
                    String cellAddress = parts[3];
                    warehouse.removeProduct(productName, quantity, cellAddress);
                } else {
                    System.out.println("Неверный формат команды.");
                }
            } else if (command.equals("INFO")) {
                warehouse.getInfo();
            } else if (command.equals("EMPTY")) {
                warehouse.showEmptyCells();
            } else {
                System.out.println("Неизвестная команда.");
            }
        }
        scanner.close();
    }
}

File: FivesLabCPP/FirstZ.cpp
//#include <iostream> // Стандартный поток ввода/вывода
//#include <map> // Стандартный ассоциативный контейнер: map
//#include <vector> // Стандартный последовательный контейнер: vector
//#include <string> // Стандартный класс строки
//#include <Windows.h> // Функции для работы с Windows, используются для установки кодовой страницы консоли
//
//using namespace std; // Использование стандартного пространства имен
//
//class Cell { // Определение класса Ячейка
//public:
//    string address; // Строка для хранения адреса ячейки
//    map<string, int> products; // Карта для хранения товаров и их количества в ячейке
//
//    Cell(string addr) : address(addr) {} // Конструктор принимает адрес в качестве параметра
//
//    void addProduct(string productName, int quantity) { // Метод для добавления товара в ячейку
//        if (products.find(productName) != products.end()) {
//            products[productName] += quantity;
//        }
//        else {
//            products[productName] = quantity;
//        }
//        cout << "Добавлено " << quantity << " единиц товара " << productName << " в ячейку " << address << "." << endl; // Вывод сообщения
//    }
//
//    void removeProduct(string productName, int quantity) { // Метод для удаления товара из ячейки
//        if (products.find(productName) != products.end() && products[productName] >= quantity) {
//            products[productName] -= quantity;
//            if (products[productName] == 0) {
//                products.erase(productName);
//            }
//            cout << "Убрано " << quantity << " единиц товара " << productName << " из ячейки " << address << "." << endl; // Вывод сообщения
//        }
//        else {
//            cout << "Невозможно убрать " << quantity << " единиц товара " << productName << " из ячейки " << address << "." << endl; // Вывод сообщения
//        }
//    }
//
//    void showContents() { // Метод для отображения содержимого ячейки
//        if (!products.empty()) {
//            cout << "Содержимое ячейки " << address << ":" << endl;
//            for (const auto& entry : products) {
//                cout << "- " << entry.first << ": " << entry.second << " шт." << endl;
//            }
//        }
//    }
//};
//
//class Warehouse { // Определение класса Склад
//public:
//    int zones; // Количество зон на складе
//    int shelvesPerZone; // Количество полок на зоне
//    int sectionsPerShelf; // Количество секций на полке
//    int shelvesCapacity; // Вместимость каждой полки
//    int totalCapacity; // Общая вместимость склада
//    map<string, Cell*> cells; // Карта для хранения ячеек на складе
//
//    Warehouse(int z, int spz, int sps, int sc) : zones(z), shelvesPerZone(spz), sectionsPerShelf(sps), shelvesCapacity(sc) { // Конструктор принимает параметры
//        totalCapacity = zones * shelvesPerZone * sectionsPerShelf * shelvesCapacity; // Вычисление общей вместимости склада
//
//        // Вложенные циклы для инициализации ячеек на складе
//        for (int zone = 1; zone <= zones; ++zone) {
//            for (int shelf = 1; shelf <= shelvesPerZone; ++shelf) {
//                for (int section = 1; section <= sectionsPerShelf; ++section) {
//                    for (int shelfPosition = 1; shelfPosition <= shelvesCapacity; ++shelfPosition) {
//                        string address = "";
//                        address += char('A' + zone - 1);
//                        address += to_string(shelf);
//                        address += to_string(section);
//                        address += to_string(shelfPosition);
//                        cells[address] = new Cell(address); // Создание новой ячейки и добавление ее в карту
//                    }
//                }
//            }
//        }
//    }
//
//    void addProduct(string productName, int quantity, string cellAddress) { // Метод для добавления товара в ячейку
//        if (cells.find(cellAddress) == cells.end()) {
//            cout << "Ячейка не существует." << endl; // Вывод сообщения, если ячейка не существует
//            return;
//        }
//        cells[cellAddress]->addProduct(productName, quantity); // Вызов метода addProduct ячейки
//    }
//
//    void removeProduct(string productName, int quantity, string cellAddress) { // Метод для удаления товара из ячейки
//        if (cells.find(cellAddress) == cells.end()) {
//            cout << "Ячейка не существует." << endl; // Вывод сообщения, если ячейка не существует
//            return;
//        }
//        cells[cellAddress]->removeProduct(productName, quantity); // Вызов метода removeProduct ячейки
//    }
//
//    void getInfo() { // Метод для получения информации о заполненных ячейках
//        vector<Cell*> filledCells; // Вектор для хранения заполненных ячеек
//        vector<string> emptyCells; // Вектор для хранения адресов пустых ячеек
//
//        int totalFilledCapacity = 0; // Общая заполненная вместимость
//        for (const auto& entry : cells) {
//            Cell* cell = entry.second;
//            if (!cell->products.empty()) {
//                filledCells.push_back(cell); // Добавление заполненной ячейки
//                totalFilledCapacity += cell->products.size(); // Обновление общей заполненной вместимости
//            }
//            else {
//                emptyCells.push_back(cell->address); // Добавление адреса пустой ячейки
//            }
//        }
//
//        double filledPercentage = (static_cast<double>(totalFilledCapacity) / totalCapacity) * 100.0; // Вычисление процента заполнения
//        cout << "Склад заполнен на " << filledPercentage << "%." << endl; // Вывод процента заполнения
//        cout << "Содержимое заполненных ячеек:" << endl;
//        for (const auto& cell : filledCells) {
//            cell->showContents(); // Отображение содержимого заполненных ячеек
//        }
//    }
//
//    void showEmptyCells() { // Метод для отображения пустых ячеек
//        cout << "Пустые ячейки:" << endl;
//        for (const auto& entry : cells) {
//            if (entry.second->products.empty()) {
//                cout << entry.first << endl; // Вывод адреса пустой ячейки
//            }
//        }
//    }
//};
//
//
//int main() {
//    setlocale(LC_ALL, "ru");
//    SetConsoleCP(1251); // установка кодовой страницы win-cp 1251 в поток ввода
//    SetConsoleOutputCP(1251); // установка кодовой страницы win-cp 1251 в поток вывода
//
//    Warehouse warehouse(2, 19, 4, 8);
//
//    cout << "Для добавления товара в ячейку используйте команду ADD <наименование товара> <количество> <адрес ячейки>." << endl;
//    cout << "Пример: ADD Апельсины 8 А1739" << endl;
//    cout << "Для удаления товара из ячейки используйте команду REMOVE <наименование товара> <количество> <адрес ячейки>." << endl;
//    cout << "Пример: REMOVE Апельсины 3 А1739" << endl;
//    cout << "Для получения информации о состоянии заполненных ячеек склада используйте команду INFO." << endl;
//    cout << "Для получения списка пустых ячеек используйте команду EMPTY." << endl;
//    cout << "Команда HELP - выводит инструкцию по вводу." << endl;
//    cout << "Команда EXIT - завершает программу." << endl;
//
//    string command; // Строка для хранения команды пользователя
//    while (true) {
//        cout << "Введите команду: "; // Подсказка для ввода команды
//        getline(cin, command); // Считывание команды пользователя из ввода
//
//        if (command == "EXIT") { // Проверка, хочет ли пользователь выйти
//            break; // Выход из цикла
//        }
//        else if (command == "HELP") { // Проверка, хочет ли пользователь получить помощь
//            // Вывод инструкций для помощи
//            cout << "Для добавления товара в ячейку используйте команду ADD <наименование товара> <количество> <адрес ячейки>." << endl;
//            cout << "Пример: ADD Апельсины 8 А1739" << endl;
//            cout << "Для удаления товара из ячейки используйте команду REMOVE <наименование товара> <количество> <адрес ячейки>." << endl;
//            cout << "Пример: REMOVE Апельсины 3 А1739" << endl;
//            cout << "Для получения информации о состоянии заполненных ячеек склада используйте команду INFO." << endl;
//            cout << "Для получения списка пустых ячеек используйте команду EMPTY." << endl;
//            cout << "Команда HELP - выводит инструкцию по вводу." << endl;
//            cout << "Команда EXIT - завершает программу." << endl;
//        }
//        else {
//            if (command.find("ADD") == 0) { // Проверка, хочет ли пользователь добавить товар
//                string productName;
//                int quantity;
//                string cellAddress;
//                cin >> productName >> quantity >> cellAddress; // Считывание названия товара, количества и адреса ячейки
//                warehouse.addProduct(productName, quantity, cellAddress); // Вызов метода addProduct склада
//            }
//            else if (command.find("REMOVE") == 0) { // Проверка, хочет ли пользователь удалить товар
//                string productName;
//                int quantity;
//                string cellAddress;
//                cin >> productName >> quantity >> cellAddress; // Считывание названия товара, количества и адреса ячейки
//                warehouse.removeProduct(productName, quantity, cellAddress); // Вызов метода removeProduct склада
//            }
//            else if (command == "INFO") { // Проверка, хочет ли пользователь получить информацию
//                warehouse.getInfo(); // Вызов метода getInfo склада
//            }
//            else if (command == "EMPTY") { // Проверка, хочет ли пользователь увидеть пустые ячейки
//                warehouse.showEmptyCells(); // Вызов метода showEmptyCells склада
//            }
//            else {
//                cout << "Неизвестная команда." << endl; // Вывод сообщения о неизвестной команде
//            }
//        }
//    }
//
//    return 0;
//}
